api/type.rb

```diff
@@ -501,36 +501,42 @@ def properties
       end
     end
 
-    # Represents an array of name=value pairs, and stores its items' type
-    class NameValues < Composite
-      # The fields which can be overridden in provider.yaml.
+    # An array of string -> string key -> value pairs, such as labels.
+    # While this is technically a map, it's split out because it's a much
+    # simpler property to generate and means we can avoid conditional logic
+    # in Map.
+    class KeyValuePairs < Composite
```

Comment:
Two questions:

1) Why is KVPairs separate from Map? Just for convenience?
2) Why doesn't KVPairs inherit from Map? Seems like we would have to do `.is_a?(Api::Type::Map) || .is_a?(Api::Type::KeyValuePairs)` instead of just `.is_a?(Api::Type::Map)`

Reply:
I split NameValues into KVPairs and Map because it felt like NameValues was two types in one, and we decided which it was with `string_to_object_map?`. That seemed like a job for a top level type instead. Every time Terraform (the only provider with any non-labels NameValues) checks for NameValues in a template, it has to check for `string_to_object_map?` as well. Examples: [#1](https://github.com/GoogleCloudPlatform/magic-modules/pull/506/files#diff-e477077e94ba4a4f3a541316723ee77b), [#2](https://github.com/GoogleCloudPlatform/magic-modules/pull/506/files#diff-49726c7e9e9080ec05710ae7d5b266df)

There is no shared code between them in Terraform, and none of the other providers have implemented a GCP resource with a map other than `labels` so there is no counterexample where code can be shared between complex maps and kv pair maps like labels. I don't see a strong reason for them to share a class when we aren't taking advantage of that relationship anywhere.

Reply:
This is fine with me. Especially with recent decisions about P/C